### PR TITLE
[FIXED JENKINS-32033] Support for multiple domain controllers

### PR DESCRIPTION
**Pull request by fbelzunc**
This PR allows to use multiple domain controllers - even with multiple domains.

@jtnord @andresrc

### CODE REVIEW

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 1]
```diff
@@ -518,7 +518,8 @@ public DirContext bind(String principalName, String password, List<SocketInfo> l
             newProps.put("java.naming.ldap.attributes.binary","tokenGroups objectSid");
             newProps.put("java.naming.ldap.factory.socket",TrustAllSocketFactory.class.getName());
             newProps.putAll(props);
-            NamingException error = null;
+            NamingException namingException = null;
+            javax.naming.AuthenticationException authenticationException = null;
```
Comment (andresrc) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: You can use `BadCredentialsException` here, as it seems the only subtype used.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 2]
```diff
@@ -537,15 +538,18 @@ public DirContext bind(String principalName, String password, List<SocketInfo> l
...
                     LOGGER.log(Level.WARNING, "Failed to bind to "+ldapServer, e);
-                    error = e; // retry
+                    namingException = e; // retry
                 }
             }
-
             // if all the attempts failed
-            throw new BadCredentialsException("Either no such user '"+principalName+"' or incorrect password", error);
+            if (authenticationException !=null ) {
+                throw new BadCredentialsException("Either no such user '" + principalName + "' or incorrect password", authenticationException);
```
Comment (andresrc) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: You can throw the previously created one if you want.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java [hunk 3]
```diff
@@ -537,15 +538,18 @@ public DirContext bind(String principalName, String password, List<SocketInfo> l
...
+            } else {
+                throw new BadCredentialsException("Either no such user '" + principalName + "' or incorrect password", namingException);
+            }
```
Comment (andresrc) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectorySecurityRealm.java]: Ditto

**Comment by jtnord** [comment]
I do not think that this was a good idea as implemented - it can lead to a user locking out their account with a single typos in the password.

**Comment by fbelzunc** [comment]
I guess that because if you fail, you will be failling consecutively on all the domain controllers, what can make the user to be locked. Is this right?

**Pull request merged**